src/codeview.jl

```diff
@@ -72,53 +72,65 @@ function is_type_unstable(code::Union{IRCode, CodeInfo}, idx::Int, used::BitSet)
 end
 
 cthulhu_warntype(args...) = cthulhu_warntype(stdout, args...)
-function cthulhu_warntype(io::IO, src, rettype, debuginfo, stable_code)
-    debuginfo = IRShow.debuginfo(debuginfo)
-    lineprinter = __debuginfo[debuginfo]
-    lambda_io::IOContext = io
-    if hasfield(typeof(src), :slotnames) && src.slotnames !== nothing
-        slotnames = Base.sourceinfo_slotnames(src)
-        lambda_io = IOContext(lambda_io, :SOURCE_SLOTNAMES => slotnames)
-        show_variables(io, src, slotnames)
+function cthulhu_warntype(io::IO, src, rettype, debuginfo, stable_code, inline_cost=false)
+    if inline_cost
+        error("Need a MethodInstance to show inlining costs. Call `cthulhu_typed` directly instead.")
     end
-    print(io, "Body")
-    InteractiveUtils.warntype_type_printer(io, rettype, true)
-    println(io)
-
-    should_print_stmt = (src isa IRCode || stable_code) ? Returns(true) : is_type_unstable
-    bb_color = (src isa IRCode && debuginfo === :compact) ? :normal : :light_black
-    irshow_config = IRShowConfig(
-        lineprinter(src), InteractiveUtils.warntype_type_printer;
-        should_print_stmt, bb_color,
-    )
-    show_ir(io, src, irshow_config)
+    cthulu_typed(io, debuginfo, src, rettype, nothing, true, stable_code, inline_cost)
     return nothing
 end
 
-
-function cthulu_typed(io::IO, debuginfo, src, rt, mi, iswarn, stable_code)
+function cthulu_typed(io::IO, debuginfo, src, rt, mi, iswarn, stable_code, inline_cost=false)
     debuginfo = IRShow.debuginfo(debuginfo)
     lineprinter = __debuginfo[debuginfo]
     rettype = ignorelimited(rt)
+    lambda_io::IOContext = io
 
     if isa(src, Core.CodeInfo)
         # we're working on pre-optimization state, need to ignore `LimitedAccuracy`
         src = copy(src)
         src.ssavaluetypes = ignorelimited.(src.ssavaluetypes::Vector{Any})
         src.rettype = ignorelimited(src.rettype)
-    end
 
-    println(io)
-    println(io, "│ ─ $(string(Callsite(-1, MICallInfo(mi, rettype), :invoke)))")
+        if src.slotnames !== nothing
+            slotnames = Base.sourceinfo_slotnames(src)
+            lambda_io = IOContext(lambda_io, :SOURCE_SLOTNAMES => slotnames)
+            iswarn && show_variables(io, src, slotnames)
+        end
+    end
 
     if iswarn
-        cthulhu_warntype(io, src, rettype, debuginfo, stable_code)
+        print(io, "Body")
+        InteractiveUtils.warntype_type_printer(io, rettype, true)
+        println(io)
     else
-        bb_color = (src isa IRCode && debuginfo === :compact) ? :normal : :light_black
-        irshow_config = IRShowConfig(lineprinter(src); bb_color)
-        show_ir(io, src, irshow_config)
+        println(io, "│ ─ $(string(Callsite(-1, MICallInfo(mi, rettype), :invoke)))")
     end
-    println(io)
+
+    if src isa IRCode && inline_cost
```

Comment:
We can just allow the inlining cost printing for unoptimized source as well ?

```suggestion
    if inline_cost
```

Reply:
I did actually experiment with that, but the issue is that any call that isn't an intrinsic is assumed to be a dynamic call, so it's probably more misleading than useful. I wonder whether we can get the recursive total costs after inlining for each call from the AbstractInterpreter, but that is probably a separate PR.

Reply:
Ah, I see.
For now maybe it'd be more kind if we print some warning message like:
```juila
        elseif toggle === :optimize
            optimize ⊻= true
            if !is_cached(mi)
                @warn "can't switch to post-optimization state, since this inference frame isn't cached"
                optimize ⊻= true
            end
            continue
        elseif toggle === :debuginfo
            debuginfo = DebugInfo((Int(debuginfo) + 1) % 3)
        elseif toggle === :inline_cost
            if !optimize
                @warn "enable optimization to see the inlining costs"
            end
            inline_cost ⊻= true
```
?

Reply:
Yeah, that sounds reasonable. Do you just want to push that to this branch?

Reply:
Okay, I will push the change and merge this.